src/AbstractMenuElement.php

```diff
@@ -0,0 +1,213 @@
+     */
+    protected function _hasChildren()
+    {
+        $children = $this->_getChildren();
+        $count    = 0;
+
+        if (is_array($children) || $children instanceof Countable) {
+            $count = count($children);
+        }
+
+        if ($children instanceof \Traversable) {
+            $count = iterator_count($children);
+        }
+
+        return $count > 0;
+    }
+
+    /**
+     * Checks whether this instance has a parent.
+     *
+     * @since [*next-version*]
+     *
+     * @return bool True if this instance has a parent; false otherwise.
+     */
+    protected function _hasParent()
+    {
+        return $this->_getParent() !== null;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateChild($child)
+    {
+        if (!$this->_isMenuElement($child)) {
+            throw $this->_createValidationFailedException('Child is not a valid menu element.', 0, null, $child);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateParent($parent)
+    {
+        if ($parent !== null && !$this->_isMenuElement($parent)) {
+            throw $this->_createValidationFailedException('Parent is not a valid menu element.', 0, null, $parent);
+        }
+    }
+
+    /**
+     * Checks if a given variable is a menu element instance.
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed $menuElement The variable to check.
+     *
+     * @return bool True if the given argument is a menu element instance, false if not.
+     */
+    protected function _isMenuElement($menuElement)
+    {
+        return $menuElement instanceof MenuElementInterface;
+    }
+
+    /**
+     * Normalizes the position.
+     *
+     * @since [*next-version*]
+     *
+     * @param int    $position The position.
+     * @param string $key      The menu key.
+     * @param string $label    The menu label.
+     *
+     * @return int
+     */
+    protected function _normalizePosition($position, $key = '', $label = '')
+    {
+        if (isset($this->children["$position"])) {
+            $hex = \md5($key . $label);
+            $dec = \base_convert($hex, 16, 10);
+
+            return $position + \substr($dec, -5) * 0.00001;
+        }
+
+        return $position;
+    }
+
+    /**
+     * Creates an exception for validation failure.
+     *
+     * @since [*next-version*]
+     *
+     * @param string                                      $message          The exception message.
+     * @param int                                         $code             The exception code.
+     * @param null                                        $inner            The previous exception in the chain.
+     * @param mixed                                       $subject          The subject that failed validation.
+     * @param string[]|StringableInterface[]|\Traversable $validationErrors The validation errors.
```

Comment:
E.g. by aliasing `StringableInterface` to `Stringable`, like I always do, and by `use`ing `Traversable`.